table_schema_cache.go

```diff
@@ -34,6 +35,57 @@ type TableSchema struct {
 	*schema.Table
 
 	CompressedColumns map[string]string // Map of column name => compression type
+
+	rowMd5Query string
+}
+
+// This query returns the MD5 hash for a row on this table. This query is valid
+// for both the source and the target shard.
+//
+// Any compressed columns specified via CompressedColumns are excluded in
+// this checksum and the raw data is returned directly.
+//
+// Note that the MD5 hash should consists of at least 1 column: the pk column.
+// This is to say that there should never be a case where the MD5 hash is
+// derived from an empty string.
+func (t *TableSchema) FingerprintQuery(schemaName, tableName string, numRows int) string {
+	columnsToSelect := make([]string, 2+len(t.CompressedColumns))
+	columnsToSelect[0] = quoteField(t.GetPKColumn(0).Name)
+	columnsToSelect[1] = t.RowMd5Query()
+	i := 2
+	for columnName, _ := range t.CompressedColumns {
+		columnsToSelect[i] = quoteField(columnName)
+		i += 1
+	}
+
+	return fmt.Sprintf(
+		"SELECT %s FROM %s WHERE %s IN (%s)",
```

Comment:
Do we need to add an explicit `ORDER BY` here?  I'm concerned if we leave it to implicit ordering it may not return deterministically for the source and target.

Reply:
All consumers of this method will read the rows and build a map of PK => fingerprints. We'll have to do this whether or not this query is done via an ORDER BY for ease of comparison (especially later with the background reverifier).

Reply:
Didn't realize it was a map 👍